File: python/01-aggregation-and-filtering/order_service.py
from order import Order
from decimal import Decimal
from collections import defaultdict


class OrderService:
    def get_total_amount_per_customer(self, orders: list[Order]) -> dict[str, Decimal]:
        if not orders:
            return {}

        valid_orders = (
            o for o in orders
            if o and o.customer and o.amount and o.currency
        )

        result = defaultdict(Decimal)

        for order in valid_orders:
            result[order.customer] += order.amount

        return dict(result)

File: javascript/01-aggregation-and-filtering/src/orderService.js
import Decimal from 'decimal.js';

export function getTotalAmountPerCustomer(orders) {

  if (orders == null) {
    return {};
  }

  return orders
    .filter(isValid)
    .reduce((acc, { customer, amount }) => {
        acc[customer] = (acc[customer] ?? new Decimal("0.0")).plus(amount);
        return acc;
    }, {})
}

function isValid(order) {
    return order
        && order.customer != null
        && order.amount != null
        && order.currency != null;
}